### PR TITLE
Call a state before block was imported.

### PR DESCRIPTION
**Pull request by tomusdrw**
and also add a (missing) test for this

### CODE REVIEW

File: core/service/src/components.rs [hunk 1]
```diff
@@ -184,6 +184,51 @@ pub trait MaintainTransactionPool<C: Components> {
...
+	PoolApi: txpool::ChainApi<Hash = Block::Hash, Block = Block>,
+{
+	use runtime_primitives::transaction_validity::TransactionValidity;
+
+	// Avoid calling into runtime if there is nothing to prune from the pool anyway.
+	if transaction_pool.status().is_empty() {
+		return Ok(())
+	}
+
+	let block = client.block(id)?;
```
Comment (rphmeier) [inline core/service/src/components.rs]: could just call header here, right?

**Comment by rphmeier** [comment]
The difference here is that it uses `parent_id` instead of the hash directly? But if you get an import notification for `X` then X has been imported -- i don't see how it was calling a state before import

**Comment by jacogr** [comment]
So I chatted to Tomasz about this one - I'm not sure about the underlying details, but can tell you my experiences from running via the API -

- submit a transaction via submitAndWtch
- get a `Ready` state back
- next state update is for `Invalid` and the node reports `invalid extrinsic`
- ... however, extrinsic gets applied, e.g. if I did a transfer the recipient balance is updated

I tested the same scenario agains this branch -

- submit a transaction via submitAndWtch
- get a `Ready` state back
- next state update is for `Finalised` with the correct hash
- extrinsic gets applied, e.g. if I did a transfer the recipient balance is updated

**Comment by tomusdrw** [comment]
@rphmeier we need to figure out the tags provided by the transaction by calling `validate_transaction` function of runtime. We can't do it after the block is imported, because such transaction now becomes invalid (so no `tags` in response). Hence we extract the tags before from the state before that block.
Previously it was done during `execute_and_import_block` phase and the tags were computed before the actual block got imported, when I moved it to a background thread I blindly copy pasted the code and didn't notice that we are actually calling a different state.

**Pull request merged**